cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,254 @@
+   *
+   * @param datasetContext the {@link DatasetContext} for getting the dataset instance.
+   * @param datasetFramework the {@link DatasetFramework} for creating the dataset instance if missing
+   * @param datasetId the {@link DatasetId} of the {@link FieldLineageDataset}
+   * @return an instance of {@link FieldLineageDataset}
+   */
+  @VisibleForTesting
+  public static FieldLineageDataset getFieldLineageDataset(DatasetContext datasetContext,
+                                                           DatasetFramework datasetFramework, DatasetId datasetId) {
+    try {
+      return DatasetsUtil.getOrCreateDataset(datasetContext, datasetFramework, datasetId,
+                                             FieldLineageDataset.class.getName(), DatasetProperties.EMPTY);
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Store the field lineage information.
+   *
+   * @param info the field lineage information
+   */
+  public void addFieldLineageInfo(FieldLineageInfo info) {
+    byte[] rowKey = getChecksumRowKey(info.getChecksum());
+    table.put(rowKey, RAW_FORMAT_COLS_BYTE, Bytes.toBytes(GSON.toJson(info)));
+  }
+
+  /**
+   * @return {@code true} if {@link FieldLineageInfo} exists in the store
+   * corresponding to the given checksum, otherwise false is returned
+   */
+  public boolean hasFieldLineageInfo(long checksum) {
+    return getFieldLineageInfo(checksum) != null;
+  }
+
+  /**
+   * Add records referring to the common operation record having the given checksum.
+   * Operations represent transformations from source endpoints to the destination endpoints.
+   * From source perspective the operations are added as forward lineage, while from
+   * destination perspective they are added as backward lineage.
+   *
+   * @param programRunId program run for which lineage is to be added
+   * @param info the FieldLineageInfo created by program run
+   */
+  public void addFieldLineageInfoReferenceRecords(ProgramRunId programRunId, FieldLineageInfo info) {
+    // For all the sources, operations represents the Forward lineage
+    for (EndPoint source : info.getSources()) {
+      addOperationReferenceRecord(programRunId, source, info.getChecksum(), FORWARD_LINEAGE_MARKER);
+    }
+
+    // For all the destinations, operations represents backward lineage
+    for (EndPoint destination : info.getDestinations()) {
+      addOperationReferenceRecord(programRunId, destination, info.getChecksum(), BACKWARD_LINEAGE_MARKER);
+    }
+  }
+
+  /**
+   * Get the field lineage information for a given endpoint within the given time range.
+   * TODO: Read methods are mainly for the testing purpose till representation from UI is finalized.
+   *
+   * @param endPoint the endpoint for which lineage information to be retrieved
+   * @param start start time in seconds
+   * @param end end time in seconds
+   * @return {@link Set} of {@link FieldLineageInfo} instances
```

Comment:
So this returns a set of FLL. Is there any way, given a FLLInfo, to know what runs produced that? Or at what time that was emitted? I am asking because one of the main use cases is triage: how did a field in a particular record get computed? If we know when it was written, or what run wrote it (people like to add "audit columns" such as the process time, or the pipeline runId to each record), can I see how that run computed this field?